File: go/pkg/memory/host.go
package memory

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"

	"github.com/haivivi/giztoy/go/pkg/embed"
	"github.com/haivivi/giztoy/go/pkg/kv"
	"github.com/haivivi/giztoy/go/pkg/recall"
	"github.com/haivivi/giztoy/go/pkg/vecstore"
)

// HostConfig configures a [Host].
type HostConfig struct {
	// Store is the shared KV store. Required.
	// The store must be created with the same Separator as configured here.
	// Each persona's data is isolated under "mem{sep}{id}{sep}..." prefixes.
	Store kv.Store

	// Vec is the shared vector index. Optional.
	// If nil, semantic vector search is disabled.
	Vec vecstore.Index

	// Embedder converts text to vectors. Optional.
	// If nil, semantic vector search is disabled.
	//
	// The embedder's Model() and Dimension() are persisted on first use.
	// Subsequent calls to NewHost with a different model or dimension will
	// return an error to prevent mixing incompatible vector spaces.
	Embedder embed.Embedder

	// Compressor is the shared [Compressor] for LLM-based conversation
	// compression. Optional. If set, all personas opened from this host
	// use it as the default compressor in [Memory.Compress].
	//
	// Callers may still pass a per-call compressor to [Memory.Compress],
	// which takes precedence over this default.
	//
	// Use [NewLLMCompressor] to create an LLM-backed compressor that
	// delegates to registered segmentors and profilers.
	Compressor Compressor

	// CompressPolicy controls when auto-compression is triggered during
	// [Conversation.Append]. If zero, [DefaultCompressPolicy] is used.
	// Set both MaxChars and MaxMessages to 0 to disable auto-compression.
	CompressPolicy CompressPolicy

	// Separator is the KV key separator byte. It must match the Store's
	// configured separator. Labels (entity labels, segment labels) must not
	// contain this character.
	//
	// Zero means [kv.DefaultSeparator] (':'), which forbids ':' in labels.
	// For natural labels like "person:小明", use a non-printable separator
	// (e.g., '\x1F') and create the KV store with the same separator.
	Separator byte
}

// embedMeta is persisted in KV to track which embedding model was used.
// On subsequent opens, the host verifies that the current Embedder matches.
type embedMeta struct {
	Model     string `json:"model"`
	Dimension int    `json:"dim"`
}

// Host is the process-level entry point for the memory system.
// It manages Memory instances for many personas, all sharing a single
// KV store, vector index, and embedder.
//
// Host is safe for concurrent use. Multiple goroutines can call Open
// simultaneously for different persona IDs.
type Host struct {
	cfg HostConfig

	mu       sync.Mutex
	memories map[string]*Memory
}

// NewHost creates a new Host and validates the embedding model consistency.
//
// If an Embedder is provided, NewHost checks the KV store for previously
// persisted model metadata. If found and the model name or dimension differs,
// NewHost returns an error. If not found, it persists the current model info.
//
// This ensures that vectors stored with one model are never searched with
// another, which would produce meaningless similarity scores.
func NewHost(ctx context.Context, cfg HostConfig) (*Host, error) {
	if cfg.Store == nil {
		return nil, fmt.Errorf("memory: HostConfig.Store is required")
	}

	if cfg.Embedder != nil {
		if err := checkEmbedMeta(ctx, cfg.Store, cfg.Embedder); err != nil {
			return nil, err
		}
	}

	return &Host{
		cfg:      cfg,
		memories: make(map[string]*Memory),
	}, nil
}

// OpenOption configures per-persona overrides when opening a [Memory].
type OpenOption func(*openConfig)

type openConfig struct {
	compressor     Compressor
	compressPolicy *CompressPolicy
	embedder       embed.Embedder
}

// WithCompressor overrides the host-level [Compressor] for this persona.
func WithCompressor(c Compressor) OpenOption {
	return func(o *openConfig) { o.compressor = c }
}

// WithCompressPolicy overrides the host-level [CompressPolicy] for this persona.
func WithCompressPolicy(p CompressPolicy) OpenOption {
	return func(o *openConfig) { o.compressPolicy = &p }
}

// WithEmbedder overrides the host-level [embed.Embedder] for this persona.
// The embedder's Dimension must match the host's configured embedder (if any)
// to ensure vector compatibility; Open returns an error on mismatch.
func WithEmbedder(e embed.Embedder) OpenOption {
	return func(o *openConfig) { o.embedder = e }
}

// Open returns a Memory for a persona. Creates the underlying recall.Index
// if this is the first call for the given ID. Subsequent calls with the
// same ID return the same Memory instance (options are ignored for already
// opened personas).
//
// The id should be a stable, unique identifier for the persona (e.g.,
// "cat_girl", "robot_boy"). It is used as the KV key prefix.
//
// Options override host-level defaults for this persona only. Use
// [WithCompressor], [WithCompressPolicy], or [WithEmbedder] to customize.
func (h *Host) Open(id string, opts ...OpenOption) (*Memory, error) {
	h.mu.Lock()
	defer h.mu.Unlock()

	if m, ok := h.memories[id]; ok {
		return m, nil
	}

	// Apply options.
	var oc openConfig
	for _, opt := range opts {
		opt(&oc)
	}

	// Resolve embedder.
	emb := h.cfg.Embedder
	if oc.embedder != nil {
		// Validate dimension compatibility.
		if emb != nil && oc.embedder.Dimension() != emb.Dimension() {
			return nil, fmt.Errorf(
				"memory: Open %q: embedder dimension mismatch: host=%d, per-persona=%d",
				id, emb.Dimension(), oc.embedder.Dimension(),
			)
		}
		emb = oc.embedder
	}

	// Resolve compressor.
	compressor := h.cfg.Compressor
	if oc.compressor != nil {
		compressor = oc.compressor
	}

	// Resolve compress policy.
	policy := h.cfg.CompressPolicy
	if oc.compressPolicy != nil {
		policy = *oc.compressPolicy
	}
	if policy.MaxChars == 0 && policy.MaxMessages == 0 && compressor != nil {
		// Apply default policy when a compressor is configured but no
		// explicit policy was set (zero value).
		policy = DefaultCompressPolicy()
	}

	idx := recall.NewIndex(recall.IndexConfig{
		Store:     h.cfg.Store,
		Embedder:  emb,
		Vec:       h.cfg.Vec,
		Prefix:    memPrefix(id),
		Separator: h.cfg.Separator,
	})

	m := newMemory(id, h.cfg.Store, idx, compressor, policy)
	h.memories[id] = m
	return m, nil
}

// Close releases all resources. After Close, the Host should not be used.
func (h *Host) Close() error {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.memories = nil
	return nil
}

// checkEmbedMeta verifies embedding model consistency. On first call it
// persists the current model; on subsequent calls it validates the stored
// model matches.
func checkEmbedMeta(ctx context.Context, store kv.Store, emb embed.Embedder) error {
	key := hostMetaKey("embed")

	current := embedMeta{
		Model:     emb.Model(),
		Dimension: emb.Dimension(),
	}

	data, err := store.Get(ctx, key)
	if err != nil {
		if err == kv.ErrNotFound {
			// First time — persist the metadata.
			return writeEmbedMeta(ctx, store, key, current)
		}
		return fmt.Errorf("memory: read embed metadata: %w", err)
	}

	// Existing metadata found — verify it matches.
	var stored embedMeta
	if err := json.Unmarshal(data, &stored); err != nil {
		return fmt.Errorf("memory: decode embed metadata: %w", err)
	}

	if stored.Model != current.Model {
		return fmt.Errorf(
			"memory: embed model mismatch: stored %q (dim=%d), current %q (dim=%d); "+
				"vectors from different models are incompatible — "+
				"either use the same model or rebuild the index",
			stored.Model, stored.Dimension, current.Model, current.Dimension,
		)
	}

	if stored.Dimension != current.Dimension {
		return fmt.Errorf(
			"memory: embed dimension mismatch for model %q: stored dim=%d, current dim=%d; "+
				"changing dimension requires rebuilding the index",
			current.Model, stored.Dimension, current.Dimension,
		)
	}

	return nil
}

func writeEmbedMeta(ctx context.Context, store kv.Store, key kv.Key, meta embedMeta) error {
	data, err := json.Marshal(meta)
	if err != nil {
		return fmt.Errorf("memory: encode embed metadata: %w", err)
	}
	if err := store.Set(ctx, key, data); err != nil {
		return fmt.Errorf("memory: write embed metadata: %w", err)
	}
	return nil
}

File: rust/memory/src/host.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use openerp_kv::KVStore;

use giztoy_embed::Embedder;
use giztoy_recall::RecallIndex;
use giztoy_vecstore::VecIndex;

use crate::error::MemoryError;
use crate::keys::{host_meta_key, mem_prefix};
use crate::memory::Memory;
use crate::types::{CompressPolicy, Compressor};

/// Configures a [Host].
pub struct HostConfig {
    /// Shared KV store. Required.
    pub store: Arc<dyn KVStore>,

    /// Shared vector index. Optional.
    pub vec: Option<Arc<dyn VecIndex>>,

    /// Converts text to vectors. Optional.
    pub embedder: Option<Arc<dyn Embedder>>,

    /// Default compressor for LLM-based compression. Optional.
    pub compressor: Option<Arc<dyn Compressor>>,

    /// Controls when auto-compression triggers.
    pub compress_policy: CompressPolicy,

    /// KV key separator byte for the graph layer. Labels must not contain
    /// this character. Default ':' forbids colon in labels. Use '\x1F'
    /// (ASCII Unit Separator) for natural labels like "person:小明".
    pub separator: char,
}

#[derive(serde::Serialize, serde::Deserialize)]
struct EmbedMeta {
    dim: usize,
}

/// Process-level entry point for the memory system.
/// Manages Memory instances for many personas, sharing a single KV store,
/// vector index, and embedder. Safe for concurrent use.
pub struct Host {
    store: Arc<dyn KVStore>,
    vec: Option<Arc<dyn VecIndex>>,
    embedder: Option<Arc<dyn Embedder>>,
    compressor: Option<Arc<dyn Compressor>>,
    policy: CompressPolicy,
    separator: char,

    memories: Mutex<HashMap<String, ()>>,
}

impl Host {
    /// Create a new Host. Validates embedding dimension consistency.
    pub fn new(cfg: HostConfig) -> Result<Self, MemoryError> {
        if let Some(ref emb) = cfg.embedder {
            check_embed_meta(cfg.store.as_ref(), emb.as_ref())?;
        }

        let mut policy = cfg.compress_policy;
        if policy.max_chars == 0 && policy.max_messages == 0 && cfg.compressor.is_some() {
            policy = CompressPolicy::default();
        }

        let separator = if cfg.separator == '\0' {
            giztoy_graph::DEFAULT_SEPARATOR
        } else {
            cfg.separator
        };

        Ok(Self {
            store: cfg.store,
            vec: cfg.vec,
            embedder: cfg.embedder,
            compressor: cfg.compressor,
            policy,
            separator,
            memories: Mutex::new(HashMap::new()),
        })
    }

    /// Open (or create) a Memory for a persona.
    pub fn open(&self, id: &str) -> Memory {
        let mut memories = self.memories.lock().expect("lock poisoned");
        memories.entry(id.to_string()).or_insert(());

        let prefix = mem_prefix(id);

        let index = RecallIndex::with_separator(
            Box::new(SharedStore(Arc::clone(&self.store))),
            Box::new(SharedStore(Arc::clone(&self.store))),
            self.embedder.as_ref().map(|e| -> Box<dyn Embedder> {
                Box::new(SharedEmbedder(Arc::clone(e)))
            }),
            self.vec.as_ref().map(|v| -> Box<dyn VecIndex> {
                Box::new(SharedVecIndex(Arc::clone(v)))
            }),
            prefix,
            self.separator,
        );

        Memory::new(
            id.to_string(),
            Arc::clone(&self.store),
            index,
            self.compressor.as_ref().map(Arc::clone),
            self.policy,
        )
    }

    /// List all persona IDs that have been opened.
    pub fn list(&self) -> Vec<String> {
        let memories = self.memories.lock().expect("lock poisoned");
        let mut ids: Vec<String> = memories.keys().cloned().collect();
        ids.sort();
        ids
    }

    /// Delete all data for a persona. Safe to call for non-existent IDs.
    pub fn delete(&self, id: &str) -> Result<(), MemoryError> {
        let persona_prefix = mem_prefix(id);
        let sep = self.separator;

        // Delete known per-persona namespaces with scoped prefix scans.
        // conv/seg/sid use ':' as namespace delimiter (fixed in key format).
        // graph uses the configurable separator, so we append it explicitly
        // to avoid matching unrelated personas (e.g. "a:goals" when deleting "a").
        let ns_prefixes = vec![
            format!("{persona_prefix}:conv:"),
            format!("{persona_prefix}:seg:"),
            format!("{persona_prefix}:sid:"),
            format!("{persona_prefix}:g{sep}"),
        ];

        for scoped in &ns_prefixes {
            let entries = self.store.scan(scoped)?;
            if !entries.is_empty() {
                let keys: Vec<&str> = entries.iter().map(|(k, _)| k.as_str()).collect();
                self.store.batch_delete(&keys)?;
            }
        }

        let mut memories = self.memories.lock().expect("lock poisoned");
        memories.remove(id);
        Ok(())
    }
}

fn check_embed_meta(store: &dyn KVStore, emb: &dyn Embedder) -> Result<(), MemoryError> {
    let key = host_meta_key("embed");
    let current = EmbedMeta { dim: emb.dimension() };

    match store.get(&key)? {
        None => {
            let data = serde_json::to_vec(&current)
                .map_err(|e| MemoryError::Serialization(e.to_string()))?;
            store.set(&key, &data)?;
            Ok(())
        }
        Some(data) => {
            let stored: EmbedMeta = serde_json::from_slice(&data)
                .map_err(|e| MemoryError::Serialization(e.to_string()))?;
            if stored.dim != current.dim {
                return Err(MemoryError::EmbedDimensionMismatch {
                    model: String::new(),
                    stored: stored.dim,
                    current: current.dim,
                });
            }
            Ok(())
        }
    }
}

// ---------------------------------------------------------------------------
// Arc wrappers: delegate trait methods to the inner Arc-shared object
// ---------------------------------------------------------------------------

struct SharedStore(Arc<dyn KVStore>);

impl KVStore for SharedStore {
    fn get(&self, key: &str) -> Result<Option<Vec<u8>>, openerp_kv::KVError> {
        self.0.get(key)
    }
    fn set(&self, key: &str, value: &[u8]) -> Result<(), openerp_kv::KVError> {
        self.0.set(key, value)
    }
    fn delete(&self, key: &str) -> Result<(), openerp_kv::KVError> {
        self.0.delete(key)
    }
    fn scan(&self, prefix: &str) -> Result<Vec<(String, Vec<u8>)>, openerp_kv::KVError> {
        self.0.scan(prefix)
    }
    fn batch_set(&self, entries: &[(&str, &[u8])]) -> Result<(), openerp_kv::KVError> {
        self.0.batch_set(entries)
    }
    fn batch_delete(&self, keys: &[&str]) -> Result<(), openerp_kv::KVError> {
        self.0.batch_delete(keys)
    }
    fn is_readonly(&self, key: &str) -> bool {
        self.0.is_readonly(key)
    }
}

struct SharedEmbedder(Arc<dyn Embedder>);

#[async_trait::async_trait]
impl Embedder for SharedEmbedder {
    async fn embed(&self, text: &str) -> Result<Vec<f32>, giztoy_embed::EmbedError> {
        self.0.embed(text).await
    }
    async fn embed_batch(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, giztoy_embed::EmbedError> {
        self.0.embed_batch(texts).await
    }
    fn dimension(&self) -> usize {
        self.0.dimension()
    }
}

struct SharedVecIndex(Arc<dyn VecIndex>);

impl VecIndex for SharedVecIndex {
    fn insert(&self, id: &str, vector: &[f32]) -> Result<(), giztoy_vecstore::VecError> {
        self.0.insert(id, vector)
    }
    fn batch_insert(&self, ids: &[&str], vectors: &[&[f32]]) -> Result<(), giztoy_vecstore::VecError> {
        self.0.batch_insert(ids, vectors)
    }
    fn search(&self, query: &[f32], top_k: usize) -> Result<Vec<giztoy_vecstore::Match>, giztoy_vecstore::VecError> {
        self.0.search(query, top_k)
    }
    fn delete(&self, id: &str) -> Result<(), giztoy_vecstore::VecError> {
        self.0.delete(id)
    }
    fn len(&self) -> usize {
        self.0.len()
    }
    fn flush(&self) -> Result<(), giztoy_vecstore::VecError> {
        self.0.flush()
    }
}
